### PR TITLE
New file utils, adjust pocl_binary.c for MSVC

### PR DESCRIPTION
**Pull request by linehill**
Also, add new test subdirectory for unit testing starting with tests on the new utils.

### CODE REVIEW

File: lib/CL/pocl_binary.c [hunk 1]
```diff
@@ -322,12 +318,12 @@ pocl_binary_get_kernel_count (cl_program program, unsigned device_i)
 /***********************************************************/
 
 /* serializes a single file. */
```
Comment (pjaaskel) [inline lib/CL/pocl_binary.c:320]: WHile at this, can you expand on the comment with the arguments described?

File: lib/CL/pocl_file_util.h [hunk 2]
```diff
@@ -24,11 +24,30 @@
...
+#include "pocl_export.h"
 
 #ifdef __cplusplus
+#include <cstdint>
 extern "C" {
+#else
+#include <stdint.h>
 #endif
 
+  typedef enum pocl_file_type_e
```
Comment (pjaaskel) [inline lib/CL/pocl_file_util.h:36]: Is the indentation wrong here due to the extern "C" braces?

File: lib/CL/pocl_file_util.h [hunk 3]
```diff
@@ -74,6 +93,49 @@ POCL_EXPORT
 int pocl_mk_tempname (char *output, const char *prefix, const char *suffix,
                       int *ret_fd);
 
+/* Gives parent directory.
```
Comment (pjaaskel) [inline lib/CL/pocl_file_util.h:96]: /**   for Doxygen to pick it up.

Then \param path ....

File: tests/unit/test_fs.cc [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Check PoCL filesystem utilities.
+
+// Copyright (c) 2024 Henry Linjamäki / Intel Finland Oy
+
```
Comment (pjaaskel) [inline tests/unit/test_fs.cc:4]: //

File: tests/unit/test_fs.cc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
```
Comment (pjaaskel) [inline tests/unit/test_fs.cc:11]: //

File: tests/unit/test_fs.cc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
```
Comment (pjaaskel) [inline tests/unit/test_fs.cc:14]: //

(just to match the existing file headers)

File: tests/unit/CMakeLists.txt [hunk 7]
```diff
@@ -42,6 +42,9 @@ function(add_unit_test MAIN_SOURCE)
 
   add_test(NAME "unit/${BASENAME}"
     COMMAND ${CMAKE_CURRENT_BINARY_DIR}/${BASENAME})
+
```
Comment (pjaaskel) [inline tests/unit/CMakeLists.txt:45]: Perhaps squash to the previous commit.

**Review by pjaaskel — CHANGES_REQUESTED** [review]

File: lib/CL/pocl_file_util.h (hunk 2 above)
Comment (linehill) [reply lib/CL/pocl_file_util.h:36]: Could be my mistake. I don't think I have run the formatter over headers.

Comment (linehill) [reply lib/CL/pocl_file_util.h:36]: According to clang-format-17, --style=GNU wants to indent the line (perhaps due to the extern "C"). I'll fix it manually.

Comment (linehill) [reply lib/CL/pocl_file_util.h]: Addressed.

File: lib/CL/pocl_binary.c (hunk 1 above)
Comment (linehill) [reply lib/CL/pocl_binary.c]: Addressed.

File: tests/unit/CMakeLists.txt (hunk 7 above)
Comment (linehill) [reply tests/unit/CMakeLists.txt:49]: Squashed now.

**Review by pjaaskel — APPROVED** [review]

**Pull request merged**